### ISSUE TITLE
Node test of the form document-node(element(foo)) seems broken

### ISSUE THREAD

**Issue by martin-honnen** [StefH/XPath2.Net#48]
It seems the implementation of the node test `document-node(element(foo))` is broken, it return all kinds of spurious results, sometimes including element nodes in the result.

For the expression `descendant-or-self::document-node(element(root))` on a document like 
``` c#
var xml = @"<root>
  <item name='item 1' value='value 1'/>
  <item name='item 2' value='value 2'/>
</root>";

XmlDocument doc = new XmlDocument();
```

the result of `doc.XPath2SelectNodes("descendant-or-self::document-node(element(root))")` contains the document node itself plus all three element nodes in the document.

**Comment by martin-honnen** [comment]
It might be because the rules for `document-node(element(foo))` inn https://github.com/StefH/XPath2.Net/blob/master/src/XPath2/XPath.y#L897 doing

```
DocumentTest
   : DOCUMENT_NODE '(' ')'
   {
      $$ = SequenceType.Document;
   }
   | DOCUMENT_NODE '(' ElementTest ')'
   {
      SequenceType type = (SequenceType)$3;
      type.TypeCode = XmlTypeCode.Document;
   }
   | DOCUMENT_NODE '(' SchemaElementTest ')'
   {
      SequenceType type = (SequenceType)$3;
      type.TypeCode = XmlTypeCode.Document;
   }
   ;
```

don't assign to `$$` as other rules do.

**Comment by martin-honnen** [comment]
Created https://github.com/martin-honnen/XPath2.Net/tree/DocumentNodeElementRuleProblem to fix this. No regresssions in the test suite and all test cases added to compare already existing `document-node()` to before failing `document-node(element(root))` are passing

**Comment by StefH** [comment]
Thanks, I'll take a look.

Note that I've added a TestRunner console app which can be used to quickly run all tests. (This is code is pushed to master)

**Comment by martin-honnen** [comment]
When I use the TestRunner with the current master branch it gives the output
```
15133 test case(s) loaded, 15133 selected.
Elapsed 00:03:48.2714349
15083 executed, 12902 (85,54%) succeeded.
12902 / 15083 = 85,54%
```
Where does the number "It conforms for 12954 from 15133 (85.60%)" on the info section come from?

**Comment by StefH** [comment]
It could be PC / OS related maybe?

I get this:
![image](https://user-images.githubusercontent.com/249938/129739720-fa8a285b-9386-4ac3-a581-bf8110c27086.png)

**Comment by StefH** [comment]
Other PC shows:
![image](https://user-images.githubusercontent.com/249938/129740598-1c408d93-929e-40bf-9471-bf78f3478b8f.png)

So maybe there is some locale / culture differences.

**Comment by martin-honnen** [comment]
I get one error or warning "Module file C:\Users\marti\Downloads\XQTS_1_0_2\TestSources\empty-lib.xq does not exists" but I guess a missing XQuery file won't affect the numbers of successfully run XPath 2 tests.

**Comment by StefH** [comment]
I did fix a threading issues with the _total and _passed.

Get the latest version.
At 1 PC I get:
`15133 executed, 12957 (85,62%) succeeded.`

At another VM I get:
`15133 executed, 12958 (85.63%) succeeded.`

The difference is 1 test:
`Test 'AdjDateTimeToTimezoneFunc/K-AdjDateTimeToTimezoneFunc-6.xqx' passed` -> this one is passed on my VM but not on my PC

This is because my VM uses UTC, but my PC not.
To solve this I used some code to set timezone to UTC for all tests.
This solves the issue.

Both runs result into:
`15133 executed, 12958 (85.63%) succeeded.`


So this should give you a stable starting point to verify if the same tests still pass if you add new functionality.
